[Review Ready] Ports baystation's modular computers

Shadowlight213 opened the pull request:
Ported from baystation12. Credit for the sprites and original code goes to them.

Mostly done, just need to finish the id card program, and do general code cleanup and finalizing.

Pictures soon, once i finish that.
Major thanks to @RemieRichards  for teaching me how to tgui and @MrStonedOne  for helping me get gulp working.

Port contains
NTNET
NTIRC
new access for network admin
Dos and revelation syndie programs
NT monitor program
File manager and peer to peer file transfer
Tablets/consoles/laptops
The various parts of those things.
Power monitor and general alarm consoles have been ported to programs

todo:
- [x] ID card program
- [x] NTIRC logging
- [x] Give power cell compatibility with chargers
- [x] mapping
- [x] give some way to make the console frames and get parts
- [x] make a circuitboard for the ntnet relay.
- [x] figure out how to fix the margins so it doesn't look like shit for some of the programs.

:cl:
add: Adds modular computers
/:cl:

I'm not integrating them into the already existing computer construction stuff because they don't use circuitboards, and work completely differently.

Shadowlight213 committed:
first attempt. partial copy and stuff.

Shadowlight213 committed:
more wip

Shadowlight213 committed:
Test

Shadowlight213 committed:
morewip

Shadowlight213 committed:
Yay tgui compiles. sorta ish

Shadowlight213 committed:
More WIP
Majority of base systems done.

Shadowlight213 committed:
Base systems pretty much done.
Still need to convert the id computer program.

Shadowlight213 committed:
Fix access.

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into bayputers

xxalpha commented:
Sounds great.

Ultimate-Chimera commented:
It's a dream come true.

AdamElTablawy commented:
Please explain what this will add/allow just in case I missed anything.

RemieRichards commented on code/modules/modular_computers/NTNet/NTNRC/conversation.dm:
```diff
@@ -0,0 +1,68 @@
+/datum/ntnet_conversation/proc/trim_message_list()
+	if(messages.len <= 50)
+		return
+	for(var/message in messages)
```
this for is better as:
```messages = messages.Copy(messages.len-50 ,0)```

RemieRichards commented on code/modules/modular_computers/NTNet/NTNRC/conversation.dm:
```diff
@@ -0,0 +1,68 @@
+		ntnet_global.chat_channels.Add(src)
+	..()
+
+/datum/ntnet_conversation/proc/add_message(var/message, var/username)
```
no ```var/``` in proc arguments

won't comment the rest.

RemieRichards commented:
gonna stop reviewing this rn cos I feel ill.

ike709 commented:
@RemieRichards Is it the flashbacks from when you worked on cool computers?

RemieRichards commented:
No it's sickness from not eating properly.

RemieRichards commented on code/modules/modular_computers/NTNet/NTNRC/conversation.dm:
```diff
@@ -0,0 +1,68 @@
+var/global/ntnrc_uid = 0
```
this should be a ```/static``` on the type.

RemieRichards commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -0,0 +1,143 @@
+	log_text += log_string
+	logs.Add(log_text)
+
+	if(logs.len > setting_maxlogcount)
```
again, not sure what it is with Bay and inefficient ways of removing old list elements, but this entire if+for is better as:

```logs = logs.Copy(logs.len-setting_maxlogcount,0)```

RemieRichards commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -0,0 +1,143 @@
+	var/operating = 0
+
+	// Check all relays. If we have at least one working relay, network is up.
+	for(var/obj/machinery/ntnet_relay/R in relays)
```
smells like an istypeless forloop.

RemieRichards commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -0,0 +1,143 @@
+	if(setting_disabled)
+		return 0
+
+	if(specific_action == NTNET_SOFTWAREDOWNLOAD)
```
a ```switch()```, atleast.

RemieRichards commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -0,0 +1,143 @@
+
+// Attempts to find a downloadable file according to filename var
+/datum/ntnet/proc/find_ntnet_file_by_name(var/filename)
+	for(var/datum/computer_file/program/P in available_station_software)
```
this should be an istypeless forloop, go through the rest as well, won't comment more.

RemieRichards commented:
Yadda yadda big PR, stopping again.

Jalleo commented on code/modules/modular_computers/NTNet/NTNRC/conversation.dm:
```diff
@@ -0,0 +1,68 @@
+	trim_message_list()
+
+/datum/ntnet_conversation/proc/trim_message_list()
+	if(messages.len <= 50)
```
Wouldn't it be better to move this check into the add message proc above?

Jalleo commented on code/modules/modular_computers/computers/item/modular_computer.dm:
```diff
@@ -0,0 +1,805 @@
+
+
+// Eject ID card from computer, if it has ID slot with card inside.
+/obj/item/modular_computer/verb/eject_id()
```
Any better way than a verb here?

Shadowlight213 replied:
Hmm. The verb is probably there so that if the computer loses power you can still get your id out.

MrStonedOne replied:
honestly, a verb still makes sense for this. uniformed way to do something that allows typed commands for power users.

Jalleo commented on code/modules/modular_computers/computers/item/modular_computer.dm:
```diff
@@ -0,0 +1,805 @@
+	proc_eject_id(usr)
+
+// Eject ID card from computer, if it has ID slot with card inside.
+/obj/item/modular_computer/verb/eject_usb()
```
Same as above.

Jalleo commented on code/modules/modular_computers/computers/item/modular_computer.dm:
```diff
@@ -0,0 +1,805 @@
+	var/issynth = issilicon(user) // Robots and AIs get different activation messages.
+	if(damage > broken_damage)
+		if(issynth)
+			user << "You send an activation signal to \the [src], but it responds with an error code. It must be damaged."
```
Fair few duplicate words here.

Jalleo commented on code/modules/modular_computers/computers/item/modular_computer.dm:
```diff
@@ -0,0 +1,805 @@
+		update_icon()
+
+
+// Checks all hardware pieces to determine if name matches, if yes, returns the hardware piece, otherwise returns null
```
This proc. Why find hardware by name?

Shadowlight213 replied:
I'm going to guess that it's so it plays nicely with the data returned from the UI.
The button uses the name of the hardware as the parameter it passes back.

xxalpha replied:
This isn't the only find_hardware_by_name proc either. Not a good idea to do it this way.

Shadowlight213 replied:
Hmm. true. Not sure what other way to do it though.

Shadowlight213 replied:
This is the only find_hardware_by_name technically.
The other procs are either child types or relayed to the machinery computer types.
I could rename the machinery version if you want I guess?

xxalpha commented on code/modules/modular_computers/file_system/programs/antagonist/dos.dm:
```diff
@@ -0,0 +1,107 @@
+/datum/computer_file/program/ntnet_dos/ui_act(action, params)
+	if(..())
+		return 1
+	world << params
```
remove this and the other world output lines

Shadowlight213 committed:
Adds header icons
Does some fixes and cleanup
Some consoles added to the map and tablets added to job loadouts for testmerging.

Shadowlight213 committed:
Removes a couple vars in arguments.

Shadowlight213 committed:
Added ntnet relay to other maps.

Shadowlight213 committed:
testmerge feedback

Shadowlight213 committed:
librarian

Shadowlight213 committed:
Fixed irc client interface

evsey9 commented:
Nice Uplink reference!

AnturK commented on code/game/machinery/recharger.dm:
```diff
@@ -127,6 +127,14 @@
 				use_power(200 * recharge_coeff)
 				using_power = 1
 
+		if(istype(charging, /obj/item/weapon/gun/energy))
```
This is probably not the path you wanted.

AnturK commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -0,0 +1,145 @@
+	for(var/F in typesof(/datum/computer_file/program))
+		var/datum/computer_file/program/prog = new F
+		// Invalid type (shouldn't be possible but just in case), invalid filetype (not executable program) or invalid filename (unset program)
+		if(!prog || !istype(prog) || prog.filename == "UnknownProgram" || prog.filetype != "PRG")
```
If you're checking the type anyway, why use typeless for

MrPerson replied:
Because F is a typepath, not an object

RemieRichards replied:
@MrPerson the ```istype()``` is still useless, as only ```/datum/computer_file/program``` can be a subtype of ```/datum/computer_file/program```, by definition. even if it was ```/obj/item/weapon/potato``` with a ```parent_type``` of ```/datum/computer_file/program```, it's -still- a program (although I don't believe ```typesof()``` will return ```parent_type``` shenanigans)

AnturK commented on code/modules/modular_computers/NTNet/NTNet_relay.dm:
```diff
@@ -0,0 +1,140 @@
+		panel_open = !panel_open
+		user << "You [panel_open ? "open" : "close"] the maintenance hatch"
+		return
+	if(istype(W, /obj/item/weapon/crowbar))
```
Don't we have default deconstruction procs on machinery for these ?

Jalleo replied:
Yes

AnturK commented on code/modules/modular_computers/computers/item/modular_computer.dm:
```diff
@@ -0,0 +1,813 @@
+		user << "It is damaged."
+
+/obj/item/modular_computer/New()
+	machines += src
```
Something probably assumes machinery is just machines. This is not the best solution.

AnturK commented on code/modules/modular_computers/computers/item/modular_computer.dm:
```diff
@@ -0,0 +1,813 @@
+	kill_program(1)
+	machines.Remove(src)
+	STOP_PROCESSING(SSmachine, src)
+	for(var/H in src.get_all_components())
```
src.

AnturK commented on code/modules/modular_computers/computers/item/modular_computer.dm:
```diff
@@ -0,0 +1,813 @@
+	return
+
+// Attempts to install the hardware into apropriate slot.
+/obj/item/modular_computer/proc/try_install_component(mob/living/user, obj/item/weapon/computer_hardware/H, found = 0)
```
These could be function of the components instead of the computer.

AnturK commented:
I'm generally against the idea of giving out everyone niche items at roundstart. And these seems like that.

Shadowlight213 commented:
@Anturk they won't get them all roundstart.
I'm only giving them all out now during testmerge for feedback and bug testing.

GunHog commented:
The RD does not have access to the NTNET diagnostic program on his office computer.

Jalleo commented:
@Anturk I think what this will allow long term is reduction of amount of consoles needed across the station by making them programmes instead.
E.G. Bridge, warden/HoS office etc.

Shadowlight213 commented:
@GunHog odd. I'll look into that.

Shadowlight213 committed:
Id computer mostly complete
Various fixes and improvements.

Militaires commented:
we really need DWAINE terminals, those things are great.

Shadowlight213 commented:
I think most people agreed that if we did that, no one would be able to figure out how to use them.
Hell, I read their source code and I still barely understand it.

Shadowlight213 committed:
Finishes id computer
adds designs for computer parts to rnd and cases to the autolathe.

Shadowlight213 committed:
Adds ntirc loggin
Adds a command console to the bridge.

Shadowlight213 commented:
Ok. That should be all the features done.
I reworked how the program authorization worked. for all programs but the ntnet monitoring one, there is no requirement to have access to run it.
However, the access is still needed to download the program from the official repository, or host it with the ptp program.
This allows you to send restricted programs to other crewmembers if you manage to steal access.

Shadowlight213 committed:
removes unneeded istype

Shadowlight213 commented:
note to self:
Fix power override when having both tesla link and battery
Make battery modules start out at full charge.

Jordie0608 commented:
note to you:
try making ID program categorized into pages to see if it reduces generate time

Shadowlight213 committed:
improvements and fixes.

Shadowlight213 commented:
Ok. I did what Jordie suggested. sorta.
Each access region has a sort of pseudo dropdown menu. you click the name of the department to make the access buttons appear. this really helps cut down on the lag since you can choose how many buttons you want visible at a time. minimizing most of the regions almost completely removes the lag.

Jordie0608 commented:
Nice

optimumtact commented:
Finish

Shadowlight213 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into bayputers

Shadowlight213 committed:
tgui rebuild

Shadowlight213 committed:
Makes try_install_component a function of the hardware.

Shadowlight213 commented:
It's pretty much done.

YakumoChen commented:
I don't see the point of this when PDAs do almost the same things

YakumoChen commented:
also I'd rather not have portable ID consoles, the immobile ones are already annoying enough All Access Machines

Shadowlight213 commented:
@YakumoChen No they don't. You're missing the point that this is a framework for datum computer programs mainly.
I can make the id computers only work with consoles if people really want that
